crates/node/Cargo.toml

```diff
@@ -128,3 +128,5 @@ sharingan = []
 sn-block-import = ["dep:mc-starknet-block-import"]
 # Enables Sierra class verification for both own and foreign blocks only + enabled for manual sealing mode
 sn-block-import-testing = ["sn-block-import"]
+# Enables dev features to improve DevX
+dev = []
```

Comment:
how can we make sure that `dev` is never allowed in production?

Reply:
Maybe with this: https://doc.rust-lang.org/stable/reference/conditional-compilation.html?highlight=debug_assertions#debug_assertions

Reply:
Do you mean we check if it's an optimized build and add an assertion that the dev flag does not exist? (like the opposite of debug_assert!). If yes, then I am not sure if that works well, because the Madara CLI will download optimized builds because they're way faster to run but it will have the dev flag enabled. I don't think there's a specific way to know that the user is running the chain in production, I think we can do

- add a big warning message whenever someone uses the Dev feature flag
- add another flag `production` and enfore that the user passes ONLY one

Reply:
A solution would me to replace all `#[cfg(feature = "debug")]` buy `#[cfg(all(debug_assertions, feature = "debug"))]` this way we will have 3 situation

- not(debug): use prod code
- debug && debug_assertions: use debug guarded code
- debug without debug_assertions (aka: --release --features=debug): can't compile


Wdyt?

Reply:
But we want `--release --features=debug` to compile for the CLI tool. We want to use release builds as debug builds are very slow but we want to enable dev features to make testing easier.

Reply:
> we want --release --features=debug

Nope, that is the very thing we want to avoid

Reply:
We have to find another way to make the CLI work, using the `not(debug)` settings

Reply:
Ya I understand that in production we don't want people to be able to run with the --release flag. However, we want to make DevX easier with Madara. It's really bad right now where you need to clone Madara, build it locally so that you can finally test it. We need built binaries that people can directly download and start testing locally. Now, handing out debug binaries ins't good because they're really slow and won't show the power of Madara. So we need a way to hand out release binaries with debug features.

Reply:
They are not that slow? Are they?

> won't show the power of Madara.

That is not what debug builds are for. They are for testing.


Test with debug binary, and prohibit dangerous features in release mode.
In the same idea I'm trying to protect against `testing` features being release out of `test` builds here: https://github.com/starkware-libs/blockifier/issues/430
I know it sounds a bit rigid, but I personally thinks it's for the greater good on the long run

Reply:
We should not expose such an easy and convenient way to fuck up your chain. People will use it.

Reply:
The DevX for fucking you chain should be as difficult as possible, so they really have to want it to do it.
Here they will do it inadvertently. 

And I put production security above the convenience of testing in the list of priorities. Especially for a blockchain project.

Reply:
I get your point but the debug build is pretty slow when you compare it to the release build. I get your point that in the current context, it's very easy to pass the dev feature flag in release mode and start your chain. 

What if we go ahead with your idea of restricting the use of `dev` feature flag in `--release` builds. But we also add another feature flag called `dev-for-release` that let's you override this restriction. This way you specifically need to enable the feature `dev-for-release` to allow this  behaviour. We would like to use this in the CLI tool but it will have a huge warning message when you use it to switch off this feature flag for production.

Reply:
> debug build is pretty slow when you compare it to the release build

How slow? And does it really matter? 
It's for local nodes and devs manually sending TXs, not for stress testing

Reply:
> How slow?

It's about 10x slower. With the release build, Madara could do 120 TPS, in debug, it maxed at 12. This is just transfers by the way. Even getting responses from the RPC is much slower. Add invoke txs is <10ms in release and 100ms in debug. 
<img width="1499" alt="Screenshot 2024-05-06 at 7 20 46 PM" src="https://github.com/keep-starknet-strange/madara/assets/95699312/ff70bc12-681a-4322-a2dd-6b25fee86e72">

> And does it really matter

It depends on the application type I guess. If I am building a game with every interaction causing a transaction (and more complicated txs than transfer), then I could be getting 2-3 TPS in debug mode which can affect my development experience. The typing game we made in Madara, for example, might have lagged here. I completely get your security concern, and it makes sense. But I feel an additional feature flag to purposefully allow dev features in production + a BIG warning message on start can solve this. wdyt?

Reply:
> But I feel an additional feature flag to purposefully allow dev features in production + a BIG warning message on start can solve this. wdyt?

To me, it's just back to step one with `feature = "dev"`

If people really need the perf for the development, they can compile to `--release` each time they modify their versionedConstants. It won't happen 100x.

Or they can compile to debug for the time they are trying different config values, and go back to `--release` when they go back to their regular dev.

But really, `--release --features=debug` this should be allowed.

Reply:
The issue is the compiling it is difficult. I am imagining users who will use madaraup or madara cli and just download the binary. They might not even have Rust installed. And increasing the step limit is a common demand for app chains.

Reply:
Hmm, cool I get your point. Let's do this then. I have pushed a commit where I check in the `build.rs` file itself and panic with a custom error message if someone uses the dev feature flag in release mode.